File: platforms/macos/Sources/OneAiRpcClient.swift
// OneAiRpcClient.swift — macOS JSON-RPC 2.0 client for `oneai app-server`
// (the sidecar transport). Replaces the old newline-JSON `OneAIBusClient`
// skeleton (which spoke the raw bus protocol to `oneai serve`) with the
// operation-oriented JSON-RPC frontend schema over a Unix domain socket.
//
// The macOS app's default transport stays in-process FFI (c_facade) — best
// UX (no process/socket). This client is the out-of-process option selected
// by the `oneai_engine_transport=sidecar` UserDefaults flag: the app spawns
// `oneai app-server --listen ipc://<ephemeral>` (see EngineProcessManager) on
// launch and routes turns/approvals/group/scenario through this client.
//
// Framing: newline-terminated JSON (matches `serve_ipc` / `serve_stdio`).
// The JSON-RPC envelope mirrors crates/oneai-app-server/src/protocol.rs
// (id: number|string on requests, no id on the `event` notification).
//
// NOT yet wired into ChatViewModel's group/scenario run path — see the
// Phase D status in the plan. The infra is complete + compilable; the live
// rewiring (and macOS runtime verification) is the remaining focused step.

import Foundation
import Network

/// One JSON-RPC `event` notification the engine emitted, decoded enough to
/// dispatch on `kind`. `params` is the full EngineYield (with its `kind` tag).
public struct OneAiEvent {
    public let kind: String
    public let params: [String: Any]
}

public protocol OneAiRpcClientDelegate: AnyObject {
    /// Called for every `event` notification. Switch on `event.kind`
    /// (turn_start / stream_chunk / thinking / tool_calls / tool_result /
    /// delegate / delegate_complete / speaker_turn / paradigm_switch /
    /// approval_request / working_state / context_accounting / plan_update /
    /// tools_added / init_result / compact_result / token_usage / error /
    /// turn_complete / iteration_start / session_* / …). Unknown kinds arrive
    /// as the bus adds yield variants (`#[non_exhaustive]`) — ignore them.
    func rpcClient(_ client: OneAiRpcClient, didReceive event: OneAiEvent)
    /// Called when the socket closed or failed. The EngineProcessManager
    /// restarts the child; the VM surfaces a transport error to the user.
    func rpcClient(_ client: OneAiRpcClient, didCloseWithError error: Error?)
}

/// A JSON-RPC 2.0 client over a Unix domain socket (the app-server ipc
/// transport). Promise-based `call(method, params)` (id correlation) + a
/// delegate for `event` notifications. Thread-safe: the pending-continuation
/// map + id counter are lock-protected; events are delivered on
/// `callbackQueue`. `@unchecked Sendable` — the lock guards all shared state
/// (NWConnection is itself thread-safe; `rxBuf` is only touched from the
/// serialized receive callback).
public final class OneAiRpcClient: @unchecked Sendable {
    private let socketPath: String
    private var connection: NWConnection?
    public weak var delegate: OneAiRpcClientDelegate?
    /// Queue the delegate callbacks fire on. Default a background queue; the
    /// VM supplies `.main` so `@Published` mutations are already on the main
    /// actor (mirrors the FFI callback path's main-thread marshalling).
    public var callbackQueue: DispatchQueue = .global(qos: .userInitiated)

    private var nextId: Int = 1
    private var pending: [Int: CheckedContinuation<[String: Any], Error>] = [:]
    private let pendingLock = NSLock()
    private var rxBuf = Data()
    private var connected = false

    public init(socketPath: String) {
        self.socketPath = socketPath
    }

    /// Connect to the app-server's UDS. Resolves once the socket is ready.
    public func connect() async throws {
        let endpoint = NWEndpoint.unix(path: socketPath)
        let conn = NWConnection(to: endpoint, using: .tcp)
        self.connection = conn
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            conn.stateUpdateHandler = { [weak self] state in
                switch state {
                case .ready:
                    self?.connected = true
                    self?.startReceiveLoop()
                    cont.resume()
                case .failed(let err):
                    cont.resume(throwing: err)
                    self?.delegate?.rpcClient(self!, didCloseWithError: err)
                case .cancelled:
                    self?.delegate?.rpcClient(self!, didCloseWithError: nil)
                default:
                    break
                }
            }
            conn.start(queue: .global(qos: .userInitiated))
        }
    }

    public func disconnect() {
        connected = false
        connection?.cancel()
        connection = nil
        // Reject anything still pending — the responses will never arrive.
        pendingLock.lock()
        let doomed = pending
        pending.removeAll()
        pendingLock.unlock()
        for (_, c) in doomed {
            c.resume(throwing: OneAiRpcError.disconnected)
        }
    }

    /// Send a JSON-RPC request; resolves with the `result` object. The id is
    /// auto-assigned. For `turn/run` the result is `{turn_id}` (resolved at
    /// TurnStart); for ack methods `{ok:true}`; `scenario/validate` returns
    /// `{ok, errors}`. Errors throw `OneAiRpcError.server(code, message)`.
    public func call(_ method: String, params: Any?) async throws -> [String: Any] {
        guard connected, let conn = connection else { throw OneAiRpcError.disconnected }
        // Assign the id + register the continuation under one lock so two
        // concurrent calls can't collide on the id counter.
        let id: Int = {
            pendingLock.lock()
            defer { pendingLock.unlock() }
            let n = nextId
            nextId += 1
            return n
        }()
        var payload: [String: Any] = ["jsonrpc": "2.0", "id": id, "method": method]
        if let params = params { payload["params"] = params }
        let data = try JSONSerialization.data(withJSONObject: payload)
        var line = String(data: data, encoding: .utf8) ?? "{}"
        if !line.hasSuffix("\n") { line.append("\n") }

        return try await withCheckedThrowingContinuation { (cont: CheckedContinuation<[String: Any], Error>) in
            self.pendingLock.lock()
            self.pending[id] = cont
            self.pendingLock.unlock()
            guard let bytes = line.data(using: .utf8) else {
                self.resolve(id, throwing: OneAiRpcError.encode)
                return
            }
            conn.send(content: bytes, completion: .contentProcessed { [weak self] err in
                if let err = err {
                    self?.resolve(id, throwing: err)
                }
            })
        }
    }

    // MARK: - Receive loop

    private func startReceiveLoop() {
        receiveChunk()
    }

    private func receiveChunk() {
        guard let conn = connection else { return }
        // UDS stream — read up to 64KiB at a time, accumulate, split on `\n`.
        conn.receive(minimumIncompleteLength: 1, maximumLength: 65_536) { [weak self] data, _, isComplete, error in
            guard let self = self else { return }
            if let error = error {
                self.handleClose(error)
                return
            }
            if let data = data {
                self.rxBuf.append(data)
                self.drainLines()
            }
            if isComplete {
                self.handleClose(nil)
                return
            }
            self.receiveChunk()
        }
    }

    private func drainLines() {
        while let nl = rxBuf.firstIndex(of: 0x0A) {
            let lineData = rxBuf.subdata(in: 0..<nl)
            rxBuf.removeSubrange(0...nl)
            guard !lineData.isEmpty,
                  let obj = try? JSONSerialization.jsonObject(with: lineData) as? [String: Any]
            else { continue }
            handleMessage(obj)
        }
    }

    private func handleMessage(_ msg: [String: Any]) {
        // Response to a pending call (has id + result/error).
        if let id = msg["id"], msg["result"] != nil || msg["error"] != nil {
            // id may arrive as Int (JSON-RPC number) or NSNumber; normalize.
            let key = (id as? Int) ?? Int((id as? NSNumber)?.intValue ?? -1)
            if key >= 0 {
                if let err = msg["error"] as? [String: Any] {
                    let code = (err["code"] as? Int) ?? -1
                    let message = (err["message"] as? String) ?? "rpc error"
                    resolve(key, throwing: OneAiRpcError.server(code: code, message: message))
                } else if let result = msg["result"] as? [String: Any] {
                    resolve(key, with: result)
                } else {
                    // result is not an object (e.g. a bare string) — wrap it.
                    resolve(key, with: ["value": msg["result"] ?? NSNull()])
                }
            }
            return
        }
        // Notification (no id). The app-server's single outbound method is
        // `event`; params is the full EngineYield with its `kind` tag.
        if msg["method"] as? String == "event", let params = msg["params"] as? [String: Any] {
            let kind = (params["kind"] as? String) ?? ""
            let ev = OneAiEvent(kind: kind, params: params)
            callbackQueue.async { [weak self] in
                guard let self = self else { return }
                self.delegate?.rpcClient(self, didReceive: ev)
            }
        }
    }

    private func resolve(_ id: Int, with result: [String: Any]) {
        pendingLock.lock()
        let cont = pending.removeValue(forKey: id)
        pendingLock.unlock()
        cont?.resume(returning: result)
    }

    private func resolve(_ id: Int, throwing error: Error) {
        pendingLock.lock()
        let cont = pending.removeValue(forKey: id)
        pendingLock.unlock()
        cont?.resume(throwing: error)
    }

    private func handleClose(_ error: Error?) {
        connected = false
        pendingLock.lock()
        let doomed = pending
        pending.removeAll()
        pendingLock.unlock()
        for (_, c) in doomed { c.resume(throwing: OneAiRpcError.disconnected) }
        callbackQueue.async { [weak self] in
            guard let self = self else { return }
            self.delegate?.rpcClient(self, didCloseWithError: error)
        }
    }
}

public enum OneAiRpcError: Error {
    case disconnected
    case encode
    case server(code: Int, message: String)
}

File: examples/native/windows/OneAiRpcClient.cs
// OneAiRpcClient.cs — Windows JSON-RPC 2.0 client for `oneai app-server`
// (the sidecar transport), over a named pipe. Mirrors the macOS
// OneAiRpcClient.swift and the VS Code server.ts: same JSON-RPC envelope,
// same id correlation, same `event` notification dispatch — only the
// transport differs (named pipe vs Unix domain socket vs stdio).
//
// The Windows app's default transport stays in-process FFI (c_facade /
// oneai_native.dll). This client is the out-of-process option: the app
// spawns `oneai app-server --listen pipe://oneai-<pid>` (see
// EngineProcessManager.cs) and routes turns/approvals/group/scenario
// through here. Per the plan, the WinUI ChatViewModel rewiring is deferred;
// this is the client + spawn infra, plus a smoke Program.cs round-trip.
//
// Framing: newline-terminated JSON (matches `serve_ipc` on Windows, which
// uses the same line framing as stdio). The pipe NAME is the flattened
// socket path (`oneai_supervisor::transport::to_pipe_name`) — pass the same
// name the app-server was started with.
//
// Windows C# is not compiled in this repo (no Windows host) — this mirrors
// the Swift/TS clients structurally and is written to be idiomatic + build
// on a Windows machine. Verify there.

using System;
using System.Collections.Concurrent;
using System.IO.Pipes;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace OneAI.Native;

/// <summary>One <c>event</c> notification the engine sent, decoded enough to
/// dispatch on <c>kind</c>.</summary>
public sealed class OneAiEvent
{
    public string Kind { get; init; } = "";
    public JsonElement Params { get; init; }
}

/// <summary>A JSON-RPC 2.0 client over a Windows named pipe (the app-server
/// ipc transport). <see cref="Call"/> is awaitable (id correlation); events
/// fire on <see cref="OnEvent"/>.</summary>
public sealed class OneAiRpcClient : IDisposable
{
    private readonly string _pipeName;
    private NamedPipeClientStream? _pipe;
    private StreamReader? _reader;
    private StreamWriter? _writer;
    private CancellationTokenSource? _cts;
    private int _nextId = 1;
    private readonly ConcurrentDictionary<int, TaskCompletionSource<JsonElement>> _pending = new();

    public event Action<OneAiEvent>? OnEvent;
    public event Action<Exception?>? OnClosed;

    /// <param name="pipeName">Pipe name without the <c>\\.\pipe\</c> prefix
    /// (e.g. <c>oneai-12345</c>), matching the name the app-server was
    /// started with via <c>--listen pipe://oneai-12345</c>.</param>
    public OneAiRpcClient(string pipeName) => _pipeName = NormalizeName(pipeName);

    private static string NormalizeName(string name)
    {
        const string prefix = @"\\.\pipe\";
        return name.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)
            ? name[prefix.Length..]
            : name;
    }

    public async Task ConnectAsync(CancellationToken ct = default)
    {
        _pipe = new NamedPipeClientStream(
            ".", _pipeName, PipeDirection.InOut, PipeOptions.Asynchronous);
        await _pipe.ConnectAsync(ct);
        _reader = new StreamReader(_pipe, new UTF8Encoding(false));
        _writer = new StreamWriter(_pipe, new UTF8Encoding(false)) { NewLine = "\n", AutoFlush = true };
        _cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        _ = Task.Run(ReceiveLoop, _cts.Token);
    }

    public void Dispose()
    {
        _cts?.Cancel();
        _writer?.Dispose();
        _reader?.Dispose();
        _pipe?.Dispose();
        // Reject anything still pending.
        foreach (var kv in _pending)
        {
            kv.Value.TrySetException(new InvalidOperationException("rpc client disposed"));
        }
        _pending.Clear();
        GC.SuppressFinalize(this);
    }

    // ── Sending requests ────────────────────────────────────────────────

    /// <summary>Send a JSON-RPC request; resolves with the <c>result</c>
    /// element. For <c>turn/run</c> that's <c>{turn_id}</c>; ack methods
    /// <c>{ok:true}</c>; <c>scenario/validate</c> <c>{ok, errors}</c>.</summary>
    public async Task<JsonElement> Call(string method, object? parameters, CancellationToken ct = default)
    {
        if (_writer is null) throw new InvalidOperationException("not connected");
        int id = Interlocked.Increment(ref _nextId);
        var tcs = new TaskCompletionSource<JsonElement>(TaskCreationOptions.RunContinuationsAsynchronously);
        _pending[id] = tcs;

        var payload = new
        {
            jsonrpc = "2.0",
            id,
            method,
            @params = parameters,
        };
        var json = JsonSerializer.Serialize(payload, payload.GetType());
        await _writer.WriteLineAsync(json.AsMemory(), ct);
        await _writer.FlushAsync(ct);

        // Reject on cancellation so the pending entry is cleaned up.
        await using var _ = ct.Register(() =>
        {
            if (_pending.TryRemove(id, out var orphan))
                orphan.TrySetCanceled(ct);
        });
        return await tcs.Task; // throws on rpc error or disconnect
    }

    private async Task ReceiveLoop()
    {
        var token = _cts!.Token;
        try
        {
            while (!token.IsCancellationRequested && _reader is not null)
            {
                var line = await _reader.ReadLineAsync(token);
                if (line is null) break; // pipe closed
                if (string.IsNullOrWhiteSpace(line)) continue;
                try
                {
                    using var doc = JsonDocument.Parse(line);
                    HandleMessage(doc.RootElement);
                }
                catch
                {
                    // Malformed line — skip (mirrors the other clients).
                }
            }
        }
        catch
        {
            // reader errored — fall through to close.
        }
        DrainPendingAndClose();
    }

    private void HandleMessage(JsonElement msg)
    {
        // Response to a pending call (has id + result/error).
        if (msg.TryGetProperty("id", out var idEl)
            && (msg.TryGetProperty("result", out _) || msg.TryGetProperty("error", out _)))
        {
            var id = idEl.GetInt32();
            if (!_pending.TryRemove(id, out var tcs)) return;
            if (msg.TryGetProperty("error", out var err))
            {
                var code = err.TryGetProperty("code", out var c) ? c.GetInt32() : -1;
                var message = err.TryGetProperty("message", out var m) ? m.GetString() ?? "rpc error" : "rpc error";
                tcs.TrySetException(new OneAiRpcException(code, message));
            }
            else if (msg.TryGetProperty("result", out var result))
            {
                tcs.TrySetResult(result.Clone());
            }
            return;
        }
        // Notification (no id) — the app-server's single outbound method `event`.
        if (msg.TryGetProperty("method", out var m) && m.GetString() == "event"
            && msg.TryGetProperty("params", out var p))
        {
            var kind = p.TryGetProperty("kind", out var k) ? k.GetString() ?? "" : "";
            OnEvent?.Invoke(new OneAiEvent { Kind = kind, Params = p.Clone() });
        }
    }

    private void DrainPendingAndClose()
    {
        foreach (var kv in _pending)
        {
            kv.Value.TrySetException(new InvalidOperationException("rpc connection closed"));
        }
        _pending.Clear();
        OnClosed?.Invoke(null);
    }
}

/// <summary>An error returned by the app-server (JSON-RPC <c>error</c>).</summary>
public sealed class OneAiRpcException : Exception
{
    public int Code { get; }
    public OneAiRpcException(int code, string message) : base($"{message} (code {code})") => Code = code;
}
